### PR TITLE
Data-layer: Add site option data-layer

### PR DESCRIPTION
**Pull request by sirreal**
Adds actions, data-layer handler and reducers for querying site options (endpoint is Jetpack only) and injecting the options into state.

This is relevant for provisioning Jetpack plans with keys.

## Questions

Is this a good approach? If the site is ever updated with default options, these values will likely be wiped out.

## Testing

Try it! The fetch action creator is untested by this flow.

Try with a Jetpack site, with Akismet key auto-provisioned by a plan (purchase a plan for a Jurassic Ninja site). While on some Calypso UI for this site, try these steps…

This option isn't included by default:

```js
state.sites.items[ state.ui.selectedSiteId ].options.wordpress_api_key
// undefined
```

Let's dispatch our action to fetch it! This endpoint was extremely slow in my testing, so watch the network tab and be patient 😞 

```js
dispatch( { type: 'SITE_OPTION_FETCH', optionName: 'wordpress_api_key', siteId: state.ui.selectedSiteId } )
```

You can check the action log:

```js
actionLog.filter(/SITE_OPTION/)
```

And observe that the option is now in state 🎉 

```js
state.sites.items[ state.ui.selectedSiteId ].options.wordpress_api_key
// your key
```

### CODE REVIEW

**Comment by sirreal** [comment]
Looks like desktop needs to enable `exportDefaultFrom`. I'll see about that…

**Comment by gwwar** [comment]
> This endpoint was extremely slow in my testing, so watch the network tab and be patient

If this will be a performance concern, let's look into optimizing this or if we need a custom endpoint before adding load.

cc @jmdodd 

Is there any need for decorating the site object vs keying by siteId? The client doesn't really need to know about server details, and I feel like the site object is quite bloated as is (we had a couple PRs a while back trying to trim down site object fields). As a more practical question, do we have enough information to display a loader?

**Comment by gwwar** [comment]
I know it was meant to be an example, but what's the `wordpress_api_key` and does the client really need it? For example, would a server to server call be a better fit? Or in the case of authentication can we do this at another layer and not expose this to Calypso?

**Comment by jsnajdr** [comment]
How will the `wordpress_api_key` option be used in the app? Will any React component retrieve the value from Redux and use it to render UI? Does the value need to be cached in Redux, or is it used only once in some check and then thrown away?

Using Calypso's most advanced data fetching mechanism by default and adding 100+ lines of code just to fetch a string value might or might not be a good idea. It depends on the use case.

Maybe a function like this does the job equally well?
```js
function checkApiKey( siteId, expectedKey ): Promise<boolean> {
  return wpcom
    .undocumented()
    .site( siteId )
    .getOption( { option_name: 'wordpress_api_key' } )
    .then( data => data.option_value === expectedKey, () => false );
}
```

**Comment by sirreal** [comment]
> Using Calypso's most advanced data fetching mechanism by default and adding 100+ lines of code just to fetch a string value might or might not be a good idea.

TL;DR I think this reality check hits the nail on the head 🙂 

I plan to pivot and use a simple approach like @jsnajdr suggests. Read on for additional replies…

The use case is porting some existing functionality that is currently part of plan setup for Jetpack sites. The flow installs, activates, and configures Akismet and VaultPress. The keys are provided as part of the plan and, I believe, they need to be set as site options.

> Is there any need for decorating the site object vs keying by siteId

There's no _need_ to put this state here, in fact there is currently an entire state tree that is dedicated –perhaps exclusively– to this flow which does contain this particular site option. I prefer to reduce duplication and put state in the most logical and least surprising place. In that sense, another site option should go with the rest of the site options. It would make sense that a selector like [`getSiteOption`](https://github.com/Automattic/wp-calypso/blob/1f9b3e6872412b47f1247d8082564afebb826bbb/client/state/sites/selectors.js#L418-L428) would find this without doing anything magical.

> The client doesn't really need to know about server details

I'm not sure what you're referring to here

> I feel like the site object is quite bloated as is (we had a couple PRs a while back trying to trim down site object fields)

Agreed. `states.sites.items[ siteId ].options` already exists, this PR is just injecting a value into it.

Whether it makes sense for site options to live as an object on the site object is another question. This may be a one-off need [better served by an alternative](https://github.com/Automattic/wp-calypso/pull/26810#issuecomment-414933358) and I'm happy to leave things as they are for now.

> I know it was meant to be an example, but what's the `wordpress_api_key` and does the client really need it?

I believe this is where the Jetpack plan provided Akismet key is stored. The client needs to know whether the correct key is provisioned. The same is true for Vaultpress with a different key/value.

> would a server to server call be a better fit

Yes, I believe so, but I'm porting existing functionality and I'd like to avoid getting too deep into the woods. I'd consider handling all of this server-side in an upcoming version.

I'll be changing directions so will close this PR. If you'd like to follow up on any questions just let me know.

**Comment by gwwar** [comment]
>The client doesn't really need to know about server details. I'm not sure what you're referring to here

Mainly that data structures don't necessarily need to match server implementations, and we can transform data as needed for client needs.

> I believe this is where the Jetpack plan provided Akismet key is stored. The client needs to know whether the correct key is provisioned. The same is true for Vaultpress with a different key/value.

Any security concerns here with exposing this key? For example, is it okay if a key is leaked to another person? It sounds like the client just needs to know if the provisioning/setup went through smoothly.

> Yes, I believe so, but I'm porting existing functionality and I'd like to avoid getting too deep into the woods. I'd consider handling all of this server-side in an upcoming version.

Depending on if we need Jetpack changes, I'd recommend queuing up this diff as soon as possible. If we think a server-side solution is more appropriate I'm not sure if it's worth the time to make an alternative flow on the client. Lannister can help with reviews if you need extra 👀 here.

**Comment by gwwar** [comment]
Thanks for the follow up @sirreal! From our slack conversation, it sounds like we want to expose the key to the user as a fallback, so it should be enough to make sure this isn't persisted in browser state 👍
